modules/service/src/main/scala/lucuma/odb/service/GuideEnvironmentService.scala

```diff
@@ -302,69 +328,193 @@ object GuideEnvironmentService {
+          val period =
+            buildAvailabilityPeriod(startTime, obsInfo, genInfo, wavelength, asterism, tracking, candidates, positions)
+          period match
+            case Left(error)                       => error.asLeft
+            case Right(period) if period.end < end => go(period.end, accum.add(period))
+            case Right(period)                     => accum.add(period.copy(end = end)).asRight
         }
-        
-      override def get(pid: Program.Id, oid: Observation.Id, obsTime: Instant)(using
+        candidates match
+          case Nil => List(AvailabilityPeriod(start, end, List.empty)).asRight
+          case _   => go(start, AvailabilityList.empty).map(_.toList)
+      }
+
+      def buildAvailabilityPeriod(
+        start:      Timestamp,
+        obsInfo:    ObservationInfo,
+        genInfo:    GeneratorInfo,
+        wavelength: Wavelength,
+        asterism:   NonEmptyList[Target],
+        tracking:   ObjectTracking,
+        candidates: List[GuideStarCandidate],
+        positions:  NonEmptyList[AgsPosition]
+      ): Either[Error, AvailabilityPeriod] =
+        for {
+          baseCoords      <- obsInfo.explicitBase
+                               .orElse(
+                                 tracking
+                                   .at(start.toInstant)
+                                   .map(_.value)
+                               )
+                               .toRight(
+                                 Error.GeneralError(
+                                   s"Unable to get coordinates for asterism in observation ${obsInfo.id}"
+                                 )
+                               )
+          scienceCoords   <- asterism.toList
+                               .traverse(t => ObjectTracking.fromTarget(t).at(start.toInstant).map(_.value))
+                               .toRight(
+                                 Error.GeneralError(
+                                   s"Unable to get coordinates for science targets in observation ${obsInfo.id}"
+                                 )
+                               )
+          scienceSpeed     = asterism.toList.map(_.masy).max
+          candidatesAt     = candidates.map(_.at(start.toInstant))
+          analysis         = Ags.agsAnalysis(obsInfo.constraints,
+                                             wavelength,
+                                             baseCoords,
+                                             scienceCoords,
+                                             positions,
+                                             genInfo.agsParams,
+                                             candidatesAt
+                             )
+          // we don't care about specific targets, just their angles and speeds.
+          // At this point, each candidate only has a single vignetting
+          candidateSpeeds  = analysis.collect { case u: AgsAnalysis.Usable => (u.vignetting.head._1, u.target.tracking.masy) }
```

Comment:
I am a bit worried about performance in dense star fields. It's something I want Andy and Bryan to look at. I `think` we could limit the results a bit more based on the observation constraintset, since we aren't caching the gaia query results like Explore is. The streaming idea is a good one. I'll have to think about it. The problem I see is that we don't know the angle before analyzing. But, we could stop when all angles are filled - which is likely to happen in a dense star field. I'll play with this for a new PR.